src/worktree.rs

```diff
@@ -0,0 +1,333 @@
+        unsafe {
+            call::c_try(raw::git_worktree_unlock(call::convert(&self.raw)))?;
+        }
+        Ok(())
+    }
+
+    /// Checks if worktree is locked
+    pub fn is_locked(&self) -> Result<WorktreeLockStatus, Error> {
+        let buf = Buf::new();
+        unsafe {
+            match try_call!(raw::git_worktree_is_locked(buf.raw(), self.raw)) {
+                0 => Ok(WorktreeLockStatus::Unlocked),
+                _ => {
+                    let v = buf.to_vec();
+                    Ok(WorktreeLockStatus::Locked(match v.len() {
+                        0 => None,
+                        _ => String::from_utf8(v).ok(),
+                    }))
+                }
+            }
+        }
+    }
+
+    /// Prunes the worktree
+    pub fn prune(&self, opts: Option<&WorktreePruneOptions>) -> Result<(), Error> {
+        // When successful the worktree should be removed however the backing structure
+        // of the git_worktree should still be valid.
+        unsafe {
+            let mut opts = opts.map(|o| o.raw());
+            try_call!(raw::git_worktree_prune(self.raw, opts.as_mut()));
+        }
+        Ok(())
+    }
+
+    /// Checks if the worktree is prunable
+    pub fn is_prunable(&self, opts: Option<&WorktreePruneOptions>) -> Result<bool, Error> {
+        unsafe {
+            let mut opts = opts.map(|o| o.raw());
+            let rv = try_call!(raw::git_worktree_is_prunable(self.raw, opts.as_mut()));
+            Ok(rv != 0)
+        }
+    }
+
+    /// Opens the repository from the worktree
+    pub fn to_repository(&self) -> Result<Repository, Error> {
+        Repository::open_from_worktree(self)
+    }
+}
+
+impl<'a> WorktreeAddOptions<'a> {
+    /// Creates a default set of add options.
+    ///
+    /// By default this will not lock the worktree
+    pub fn new() -> WorktreeAddOptions<'a> {
+        WorktreeAddOptions {
+            lock: false,
+            reference: None,
+        }
+    }
+
+    /// If enabled, this will cause the newly added worktree to be locked
+    pub fn lock(&mut self, enabled: bool) -> &mut WorktreeAddOptions<'a> {
+        self.lock = enabled;
+        self
+    }
+
+    /// reference to use for the new worktree HEAD
+    pub fn reference(&mut self, reference: Option<Reference<'a>>) -> &mut WorktreeAddOptions<'a> {
+        self.reference = reference;
+        self
+    }
+
+    /// Creates a set of raw add options to be used with `git_worktree_add`
+    ///
+    /// This method is unsafe as the returned value may have pointers to the
+    /// interior of this structure
+    pub unsafe fn raw(&self) -> raw::git_worktree_add_options {
+        let mut opts = mem::zeroed();
+        assert_eq!(
+            raw::git_worktree_add_options_init(&mut opts, raw::GIT_WORKTREE_ADD_OPTIONS_VERSION),
+            0
+        );
+
+        opts.lock = self.lock as c_int;
+        opts.reference = crate::call::convert(&self.reference.as_ref().map(|o| o.raw()));
```

Comment:
Instead of the `convert` function could this do the conversion manually?

Also I think it might be good to store the `git_worktree_add_options` directly in this structure.